build_tools/ci/matmul_ukernel_vs_codegen_benchmark/test.cpp

```diff
@@ -0,0 +1,261 @@
+        return 1;
+      }
+      instr_v.push_back(a);
+    }
+  }
+
+  // Get kernel from xclbin and register it with a device:
+  auto device = xrt::device(/* device_index */ 0);
+  auto xkernels = xclbin.get_kernels();
+  if (xkernels.size() != 1) {
+    std::cerr << "Kernel count: " << xkernels.size()
+              << ". Kernel count must be 1" << std::endl;
+    return 1;
+  }
+  auto xkernel = xkernels[0];
+  auto kernelName = xkernel.get_name();
+  device.register_xclbin(xclbin);
+
+  // Parse M, K, N:
+  int M, K, N;
+  std::string MStr = argv[3];
+  M = std::stoi(MStr);
+  std::string KStr = argv[4];
+  K = std::stoi(KStr);
+  std::string NStr = argv[5];
+  N = std::stoi(NStr);
+
+
+  // Obtain results file (full path) from argv[6]:
+  std::string results_file = argv[6];
+
+  // Check the results_file is a valid file, and we can open it:
+  {
+    std::ofstream results(results_file);
+    if (!results.good()) {
+      std::cerr << "Unable to open results file: " << results_file << std::endl;
+      return 1;
+    }
+  }
+
+
+  xrt::hw_context context(device, xclbin.get_uuid());
+  auto kernel = xrt::kernel(context, kernelName);
+
+  int sizeInstructions = instr_v.size() * sizeof(int32_t);
+  int nBytesA = M * K * sizeof(bf16);
+  int nBytesB = N * K * sizeof(bf16);
+  int nBytesC = M * N * sizeof(float);
+
+  auto bo_instr = xrt::bo(device, sizeInstructions, XCL_BO_FLAGS_CACHEABLE,
+                          kernel.group_id(0));
+  auto bo_a =
+      xrt::bo(device, nBytesA, XRT_BO_FLAGS_HOST_ONLY, kernel.group_id(2));
+  auto bo_b =
+      xrt::bo(device, nBytesB, XRT_BO_FLAGS_HOST_ONLY, kernel.group_id(3));
+  auto bo_c =
+      xrt::bo(device, nBytesC, XRT_BO_FLAGS_HOST_ONLY, kernel.group_id(4));
+
+  void *bufInstr = bo_instr.map<void *>();
+  memcpy(bufInstr, instr_v.data(), instr_v.size() * sizeof(int32_t));
+
+  auto aFloatValues = generateRandomFloats(M * K, 0);
+  auto bFloatValues = generateRandomFloats(N * K, 1);
+
+  std::vector<bf16> aBfValues(M * K);
+  std::vector<bf16> bBfValues(N * K);
+  for (int i = 0; i < M * K; ++i) {
+    aBfValues[i] = float_to_bf16(aFloatValues[i]);
+  }
+  for (int i = 0; i < N * K; ++i) {
+    bBfValues[i] = float_to_bf16(bFloatValues[i]);
+  }
+
+  // Initialize A and B with 1's:
+  bf16 *bufA = bo_a.map<bf16 *>();
+  memcpy(bufA, aBfValues.data(), (aBfValues.size() * sizeof(bf16)));
+
+  bf16 *bufB = bo_b.map<bf16 *>();
+  memcpy(bufB, bBfValues.data(), (bBfValues.size() * sizeof(bf16)));
+
+  float *bufC = bo_c.map<float *>();
+
+  int nIters = 10;
+
+  std::vector<double> allTimes;
+  for (int i = 0; i < nIters; ++i) {
+
+    bo_instr.sync(XCL_BO_SYNC_BO_TO_DEVICE);
+    bo_a.sync(XCL_BO_SYNC_BO_TO_DEVICE);
+    bo_b.sync(XCL_BO_SYNC_BO_TO_DEVICE);
+    bo_c.sync(XCL_BO_SYNC_BO_TO_DEVICE);
+
+    // We will time the run, and print it.
+    auto start = std::chrono::high_resolution_clock::now();
+    auto run = kernel(bo_instr, instr_v.size(), bo_a, bo_b, bo_c);
```

Comment:
was this kernel generated with latest iree-amd-aie, if the answer is yes then this would be the wrong way to run it. You can copy the changes done in the changes done by https://github.com/Xilinx/mlir-aie/pull/1517